File: typescript/tracing/writer/04_tool_calling.ts
import {
  createRespan,
  createWriterClient,
  DEFAULT_CHAT_MODEL,
  logExampleResult,
  runWithWriterWorkflow,
  shutdownRespan,
  withTimeout,
} from "./_shared.js";

const workflowName = "writer.tool_calling.workflow";
const respan = createRespan();
const writer = createWriterClient();

function getWeather(args: { city: string }): { city: string; forecast: string; temperature_c: number } {
  return { city: args.city, forecast: "clear", temperature_c: 23 };
}

try {
  const finalCompletion = await withTimeout(
    runWithWriterWorkflow(respan, workflowName, async () => {
      const first = await writer.chat.chat({
        model: DEFAULT_CHAT_MODEL,
        messages: [{ role: "user", content: "Use a tool to check the weather in Tokyo." }],
        tools: [
          {
            type: "function",
            function: {
              name: "get_weather",
              description: "Get the current weather for a city.",
              parameters: {
                type: "object",
                properties: { city: { type: "string" } },
                required: ["city"],
              },
            },
          },
        ],
        tool_choice: { value: "auto" },
      });

      const toolCall = first.choices[0]?.message?.tool_calls?.[0];
      if (!toolCall) {
        throw new Error("Writer did not return a tool call.");
      }

      const args = JSON.parse(toolCall.function.arguments || "{}") as { city: string };
      const toolResult = getWeather(args);

      return await writer.chat.chat({
        model: DEFAULT_CHAT_MODEL,
        messages: [
          { role: "user", content: "Use a tool to check the weather in Tokyo." },
          first.choices[0].message,
          {
            role: "tool",
            tool_call_id: toolCall.id,
            content: JSON.stringify(toolResult),
          },
        ],
      });
    }),
    workflowName,
  );

  logExampleResult(workflowName, {
    expected: "two chat spans plus a tool execution child span",
    actual: finalCompletion.choices[0]?.message?.content,
  });
} finally {
  await shutdownRespan(respan);
}

File: python/tracing/portkey/04_tool_calling.py
from __future__ import annotations

import json

from _shared import (
    example_attributes,
    execution_id,
    finish_respan,
    make_client,
    make_respan,
    marker,
    model_name,
    print_result,
    workflow_name,
)
from respan import tool, workflow

EXAMPLE_NAME = "tool-calling"


@tool(name="get_weather")
def get_weather(city: str) -> str:
    return f"{city} is sunny and 72F."


@workflow(name=workflow_name(EXAMPLE_NAME))
def trace_tool(city: str) -> dict[str, str]:
    client = make_client()
    messages: list[dict[str, object]] = [
        {"role": "user", "content": f"What is the weather in {city}?"}
    ]
    tools = [
        {
            "type": "function",
            "function": {
                "name": "get_weather",
                "description": "Return deterministic weather.",
                "parameters": {
                    "type": "object",
                    "properties": {"city": {"type": "string"}},
                    "required": ["city"],
                },
            },
        }
    ]
    try:
        first = client.chat.completions.create(
            model=model_name(), messages=messages, tools=tools, tool_choice="auto"
        )
        assistant = first.choices[0].message
        call = assistant.tool_calls[0]
        arguments = json.loads(call.function.arguments)
        result = get_weather(arguments["city"])
        messages.extend(
            [
                assistant.model_dump(exclude_none=True),
                {"role": "tool", "tool_call_id": call.id, "content": result},
            ]
        )
        second = client.chat.completions.create(
            model=model_name(), messages=messages, tools=tools
        )
        return {
            "tool_result": result,
            "response": second.choices[0].message.content or "",
        }
    finally:
        client.close()


def main() -> None:
    run_marker = marker()
    execution = execution_id()
    respan = make_respan(EXAMPLE_NAME, run_marker)
    try:
        with example_attributes(
            EXAMPLE_NAME, run_marker, execution, mode="deterministic"
        ):
            result = trace_tool("Tokyo")
        print_result(EXAMPLE_NAME, run_marker, result)
    finally:
        finish_respan(respan)


if __name__ == "__main__":
    main()
